Flaky of TestModelRouteWithRateLimit

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
-->

**What happened**:

E2E Flaky of https://github.com/volcano-sh/kthena/actions/runs/21200380737/job/60984687072?pr=649

```bash
=== RUN   TestModelRouteWithRateLimit
=== RUN   TestModelRouteWithRateLimit/VerifyInputTokenRateLimitEnforcement
    shared.go:474: Test 1: Verifying input token rate limit
    shared.go:509: 
        	Error Trace:	/home/runner/work/kthena/kthena/test/e2e/router/shared.go:509
        	Error:      	Not equal: 
        	            	expected: 200
        	            	actual  : 404
        	Test:       	TestModelRouteWithRateLimit/VerifyInputTokenRateLimitEnforcement
        	Messages:   	Request 1 should succeed (consumed ~10/30 tokens). Response: "request to all pods failed""request processing failed"
=== RUN   TestModelRouteWithRateLimit/VerifyRateLimitWindowAccuracy
    shared.go:532: Test 2: Verifying rate limit window accuracy...
    shared.go:568: Waiting 10s (within rate limit window)...
    shared.go:578: Waiting additional 1m0s for window reset (total: 1m10s)...
    shared.go:587:  Rate limit window accuracy verified
=== RUN   TestModelRouteWithRateLimit/VerifyRateLimitResetMechanism
    shared.go:592: Test 3: Verifying rate limit reset mechanism...
    shared.go:630: Waiting 1m10s for complete rate limit window reset...
    shared.go:647: Rate limit reset mechanism verified (quota restored: 3 requests)
=== RUN   TestModelRouteWithRateLimit/VerifyOutputTokenRateLimitEnforcement
    shared.go:652: Test 4: Verifying output token rate limit (100 tokens/minute)...
    shared.go:703: Request 1 succeeded, response size: 392 bytes (total: 392 bytes)
    shared.go:706: Output rate limited after 1 requests
    shared.go:721:  Output token rate limit enforced after 1 requests
--- FAIL: TestModelRouteWithRateLimit (142.12s)
    --- FAIL: TestModelRouteWithRateLimit/VerifyInputTokenRateLimitEnforcement (0.02s)
    --- PASS: TestModelRouteWithRateLimit/VerifyRateLimitWindowAccuracy (70.04s)
    --- PASS: TestModelRouteWithRateLimit/VerifyRateLimitResetMechanism (70.04s)
    --- PASS: TestModelRouteWithRateLimit/VerifyOutputTokenRateLimitEnforcement (2.02s)
```

@katara-Jayprakash As I asked you many times, why you use your own `SendChatRequest` instead of `CheckChatCompletions` which would retry enough times. Because Router need time to reconcile resources like ModelRoute/ModelServer.

Please fix this issue.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:

- Kthena version:
- Kubernetes version:
- Others:

--- a/test/e2e/router/shared.go
+++ b/test/e2e/router/shared.go
@@ -492,38 +492,39 @@ func TestModelRouteWithRateLimitShared(t *testing.T, testCtx *routercontext.Rout
 			return err == nil && mr != nil
 		}, 2*time.Minute, 2*time.Second, "ModelRoute should be created")
 
-		// Send requests until rate limited
-		expectedSuccessfulRequests := inputTokenLimit / tokensPerRequest
-		for i := 0; i < expectedSuccessfulRequests; i++ {
-			if i == 0 {
-				//  use CheckChatCompletions (handles router reconciliation with retries)
-				utils.CheckChatCompletions(t, createdModelRoute.Spec.ModelName, standardMessage)
-				t.Logf("Request %d succeeded (with retry logic for reconciliation)", i+1)
-			} else {
-				// Subsequent requests: use SendChatRequest
-				resp := utils.SendChatRequest(t, createdModelRoute.Spec.ModelName, standardMessage)
-				responseBody, readErr := io.ReadAll(resp.Body)
-				resp.Body.Close()
+		// First request: use CheckChatCompletions to handle router reconciliation
+		resp := utils.CheckChatCompletions(t, createdModelRoute.Spec.ModelName, standardMessage)
+		tokensConsumed := resp.Attempts * tokensPerRequest
+		t.Logf("Router reconciliation complete (consumed %d tokens in %d attempts)", tokensConsumed, resp.Attempts)
 
-				require.NoError(t, readErr, "Failed to read response body on request %d", i+1)
-				require.Equal(t, http.StatusOK, resp.StatusCode,
-					"Request %d should succeed. Response: %s", i+1, string(responseBody))
-				t.Logf("Request %d succeeded", i+1)
-			}
+		// Calculate remaining quota
+		remainingQuota := inputTokenLimit - tokensConsumed
+		expectedSuccessfulRequests := remainingQuota / tokensPerRequest
+
+		// Send remaining requests until quota exhausted
+		for i := 0; i < expectedSuccessfulRequests; i++ {
+			resp := utils.SendChatRequest(t, createdModelRoute.Spec.ModelName, standardMessage)
+			responseBody, readErr := io.ReadAll(resp.Body)
+			resp.Body.Close()
+
+			require.NoError(t, readErr, "Failed to read response body on request %d", i+1)
+			require.Equal(t, http.StatusOK, resp.StatusCode,
+				"Request %d should succeed. Response: %s", i+1, string(responseBody))
+			t.Logf("Request %d succeeded", i+1)
 		}
 
-		// Next request should be rate limited (quota exhausted)
+		// Next request should be rate limited
 		rateLimitedResp := utils.SendChatRequest(t, createdModelRoute.Spec.ModelName, standardMessage)
 		responseBody, readErr := io.ReadAll(rateLimitedResp.Body)
 		rateLimitedResp.Body.Close()
 
 		require.NoError(t, readErr, "Failed to read rate limit response body")
 		assert.Equal(t, http.StatusTooManyRequests, rateLimitedResp.StatusCode,
-			"Request %d should be rate limited", expectedSuccessfulRequests+1)
+			"Request should be rate limited after exhausting quota")
 		assert.Contains(t, strings.ToLower(string(responseBody)), "rate limit",
 			"Rate limit error response must contain descriptive message")
 
-		t.Logf("Input token rate limit enforced after %d requests", expectedSuccessfulRequests)
+		t.Logf("Input token rate limit enforced after %d total requests", resp.Attempts+expectedSuccessfulRequests)
 	})
 
 	// Test 2 Verify rate limit window accuracy and persistence
@@ -549,17 +550,17 @@ func TestModelRouteWithRateLimitShared(t *testing.T, testCtx *routercontext.Rout
 			return err == nil && mr != nil
 		}, 2*time.Minute, 2*time.Second, "ModelRoute should be created")
 
-		// Exhaust quota to ensure rate limit is active
-		expectedSuccessfulRequests := inputTokenLimit / tokensPerRequest
+		// First request: handle reconciliation and track tokens
+		resp := utils.CheckChatCompletions(t, createdModelRoute.Spec.ModelName, standardMessage)
+		tokensConsumed := resp.Attempts * tokensPerRequest
+
+		// Exhaust remaining quota
+		remainingQuota := inputTokenLimit - tokensConsumed
+		expectedSuccessfulRequests := remainingQuota / tokensPerRequest
 		for i := 0; i < expectedSuccessfulRequests; i++ {
-			if i == 0 {
-				//  CheckChatCompletions for handling  router reconciliation
-				utils.CheckChatCompletions(t, createdModelRoute.Spec.ModelName, standardMessage)
-			} else {
-				resp := utils.SendChatRequest(t, createdModelRoute.Spec.ModelName, standardMessage)
-				resp.Body.Close()
-				assert.Equal(t, http.StatusOK, resp.StatusCode, "Request %d should succeed", i+1)
-			}
+			resp := utils.SendChatRequest(t, createdModelRoute.Spec.ModelName, standardMessage)
+			resp.Body.Close()
+			assert.Equal(t, http.StatusOK, resp.StatusCode, "Request %d should succeed", i+1)
 		}
 
 		// Verify rate limit is active
@@ -614,17 +615,18 @@ func TestModelRouteWithRateLimitShared(t *testing.T, testCtx *routercontext.Rout
 			return err == nil && mr != nil
 		}, 2*time.Minute, 2*time.Second, "ModelRoute should be created")
 
-		// Consume the quota
-		expectedSuccessfulRequests := inputTokenLimit / tokensPerRequest
+		// First request: handle reconciliation and track tokens
+		resp := utils.CheckChatCompletions(t, createdModelRoute.Spec.ModelName, standardMessage)
+		tokensConsumed := resp.Attempts * tokensPerRequest
+
+		// Consume remaining quota
+		remainingQuota := inputTokenLimit - tokensConsumed
+		expectedSuccessfulRequests := remainingQuota / tokensPerRequest
 		for i := 0; i < expectedSuccessfulRequests; i++ {
-			if i == 0 {
-				utils.CheckChatCompletions(t, createdModelRoute.Spec.ModelName, standardMessage)
-			} else {
-				resp := utils.SendChatRequest(t, createdModelRoute.Spec.ModelName, standardMessage)
-				resp.Body.Close()
-				assert.Equal(t, http.StatusOK, resp.StatusCode,
-					"Request %d should succeed", i+1)
-			}
+			resp := utils.SendChatRequest(t, createdModelRoute.Spec.ModelName, standardMessage)
+			resp.Body.Close()
+			assert.Equal(t, http.StatusOK, resp.StatusCode,
+				"Request %d should succeed", i+1)
 		}
 
 		// Confirm rate limiting is active

--- a/test/e2e/utils/chat.go
+++ b/test/e2e/utils/chat.go
@@ -52,6 +52,7 @@ type ChatCompletionsRequest struct {
 type ChatCompletionsResponse struct {
 	StatusCode int
 	Body       string
+	Attempts   int
 }
 
 // CheckChatCompletions sends a chat completions request to the router service and verifies the response.
@@ -95,8 +96,9 @@ func SendChatRequestWithRetry(t *testing.T, url string, modelName string, messag
 
 	var resp *http.Response
 	var responseStr string
+	var attempt int
 
-	for attempt := 0; attempt < maxRetries; attempt++ {
+	for attempt = 0; attempt < maxRetries; attempt++ {
 		req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
 		require.NoError(t, err, "Failed to create HTTP request")
 		req.Header.Set("Content-Type", "application/json")
@@ -156,6 +158,7 @@ func SendChatRequestWithRetry(t *testing.T, url string, modelName string, messag
 	return &ChatCompletionsResponse{
 		StatusCode: resp.StatusCode,
 		Body:       responseStr,
+		Attempts:   attempt + 1,
 	}
 }
 

